File: python/Fruits_Segmentation.py
# export results to excel file
from openpyxl import load_workbook
import xlsxwriter
# using OpenCV library to processing image
import cv2
import numpy as np
# estimate time processing
from timeit import default_timer as timer
import time
# path
import os
import concurrent.futures

print("[INFOR]: FRUITS SEGMENTATION.")

InputFileName_Idx = 0

def multiThread(fileNames_list):
    """
    """
    with concurrent.futures.ThreadPoolExecutor() as executor:
        executor.map(fruits_feature_extraction, fileNames_list)

def fruits_feature_extraction(fileName):
    global InputFileName_Idx
    global InputFolder_Path_str
    global OutputFolder_GrayImage_Path_str
    global OutputFolder_BlurImage_Path_str
    global OutputFolder_InrangeImage_Path_str
    global OutputFolder_CannyImage_Path_str
    global OutputFolder_DilateImage_Path_str
    global OutputFolder_FloodFillImage_Path_str
    global OutputFolder_RemoveNoiseImage_Path_str
    global OutputFolder_ContourImage_Path_str
    global OutputFolder_AreaImage_Path_str
    global OutputFolder_IsolateImage_Path_str
    global OutputFolder_DimensionImage_Path_str

    # load image
    Original_Image = cv2.imread(InputFolder_Path_str + fileName)

    if ImageProcessing_Method_str == "hsv":
        # convert BGR to HSV
        HSV_Image = cv2.cvtColor(Original_Image, cv2.COLOR_BGR2HSV)
        if SaveImageProcessingResult_Check_bool:
            # save image
            cv2.imwrite(OutputFolder_HsvImage_Path_str + fileName, HSV_Image)
        # convert to grayscale
        GrayScale_Image = cv2.cvtColor(HSV_Image, cv2.COLOR_BGR2GRAY)
        if SaveImageProcessingResult_Check_bool:
            # save image
            cv2.imwrite(OutputFolder_GrayImage_Path_str + fileName, GrayScale_Image)
    else:
        # convert to grayscale
        GrayScale_Image = cv2.cvtColor(Original_Image, cv2.COLOR_BGR2GRAY)
        if SaveImageProcessingResult_Check_bool:
            # save image
            cv2.imwrite(OutputFolder_GrayImage_Path_str + fileName, GrayScale_Image)

    M = GrayScale_Image.shape[0] # Height of image
    N = GrayScale_Image.shape[1] # Width of image

    # smooth image and reduce noise of background
    Blur_Image = cv2.blur(GrayScale_Image, Blur_KernelSize_tuple)
    if SaveImageProcessingResult_Check_bool:
        # save image
        cv2.imwrite(OutputFolder_BlurImage_Path_str + fileName, Blur_Image)

    if ImageProcessing_Method_str == "inrange":
        # inRange
        inRange_Image = cv2.inRange(Blur_Image, inRange1_int, inRange2_int)
        if SaveImageProcessingResult_Check_bool:
            # save image
            cv2.imwrite(OutputFolder_InrangeImage_Path_str + fileName, inRange_Image)

        # using canny algorithm to find out the edge
        Canny_Image = cv2.Canny(inRange_Image, ThresholdCanny1_int, ThresholdCanny2_int)
        if SaveImageProcessingResult_Check_bool:
            # save image
            cv2.imwrite(OutputFolder_CannyImage_Path_str + fileName, Canny_Image)
    else:
        # using canny algorithm to find out the edge
        Canny_Image = cv2.Canny(Blur_Image, ThresholdCanny1_int, ThresholdCanny2_int)
        if SaveImageProcessingResult_Check_bool:
            # save image
            cv2.imwrite(OutputFolder_CannyImage_Path_str + fileName, Canny_Image)

    # using dilate operation to try to find connected edge components
    Dilate_Image = cv2.dilate(Canny_Image, Dilate_KernelSize_np, iterations=1)
    if SaveImageProcessingResult_Check_bool:
        # save image
        cv2.imwrite(OutputFolder_DilateImage_Path_str + fileName, Dilate_Image)

    # using FloodFill to remove noise
    FloodFill_Image = Dilate_Image
    Gray_FloodFill = Gray_FloodFill_Ini_int
    Mask_FloodFill = np.zeros((M+2, N+2), np.uint8)
    for x in range(0, M):
        for y in range(0, N):
            r = Dilate_Image[x, y]
            if (r == L - 1):
                cv2.floodFill(FloodFill_Image, Mask_FloodFill, (y, x), Gray_FloodFill)
                Gray_FloodFill += Gray_FloodFill_Step_int
    if SaveImageProcessingResult_Check_bool:
        # save image
        cv2.imwrite(OutputFolder_FloodFillImage_Path_str + fileName, FloodFill_Image)
    FloodFill_AreaComponents_list = [0] * L
    for x in range(0, M):
        for y in range(0, N):
            r = FloodFill_Image[x, y]
            if (r > 0):
                FloodFill_AreaComponents_list[r] += 1
    FloodFill_AreaComponents_Enumerate = list(enumerate(FloodFill_AreaComponents_list))
    # sorting 'FloodFill_AreaComponents_Enumerate' from max to min
    FloodFill_AreaComponents_MaxMin = sorted(FloodFill_AreaComponents_Enumerate, key=lambda x:x [1], reverse = True)
    rmax = FloodFill_AreaComponents_MaxMin[FloodFill_AreaComponents_MaxMin_Idx0_int][FloodFill_AreaComponents_MaxMin_Idx1_int]
    RemoveNoise_Image = np.zeros((GrayScale_Image.shape[0], GrayScale_Image.shape[1], 1), np.uint8)
    for x in range(0, M):
        for y in range(0, N):
            r = FloodFill_Image[x, y]
            if (r == rmax):
                RemoveNoise_Image[x, y] = L - 1
            else:
                RemoveNoise_Image[x, y] = 0
    if SaveImageProcessingResult_Check_bool:
        # save image
        cv2.imwrite(OutputFolder_RemoveNoiseImage_Path_str + fileName, RemoveNoise_Image)

    # find contours
    """
    Because canny edge detector separate different segments,
    so, using 'findContours' to assemble those edge pixels.
    NOTE: The function 'findContours' computes contours from 
            binary images, so it take images created by 'Canny'
    """
    _, Contours, Hierarchy = cv2.findContours(RemoveNoise_Image, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    # find fruit contour
    Number_of_Contours = len(Contours)
    # using bubble sort to sorting contours from max to min
    Contours_Sorting_MaxMin = Contours.copy()
    for i in range(Number_of_Contours):
        for j in range(0, Number_of_Contours - i - 1):
            if (len(Contours_Sorting_MaxMin[j]) < len(Contours_Sorting_MaxMin[j + 1])):
                Contours_Sorting_MaxMin[j], Contours_Sorting_MaxMin[j + 1] = Contours_Sorting_MaxMin[j + 1], Contours_Sorting_MaxMin[j]
    # draw fruit contour '0'
    DrawFruitContour_Image = np.zeros((GrayScale_Image.shape[0], GrayScale_Image.shape[1], 1), np.uint8)
    cv2.drawContours(DrawFruitContour_Image, Contours_Sorting_MaxMin, Fruit_Contours_SortingMaxMin_Idx_int, (255, 0, 0), 1)
    if SaveImageProcessingResult_Check_bool:
        # save image
        cv2.imwrite(OutputFolder_ContourImage_Path_str + fileName, DrawFruitContour_Image)

    # Measure fruit dimensions
    # drawing rectangle around contour
    Dimensions_Image = DrawFruitContour_Image
    (X_Fruit_Contour, Y_Fruit_Contour, W_Fruit_Contour, H_Fruit_Contour) = cv2.boundingRect(Contours_Sorting_MaxMin[Fruit_Contours_SortingMaxMin_Idx_int])
    cv2.rectangle(Dimensions_Image, (X_Fruit_Contour, Y_Fruit_Contour), (X_Fruit_Contour + W_Fruit_Contour, Y_Fruit_Contour + H_Fruit_Contour), (255, 0, 0), 1)
    if SaveImageProcessingResult_Check_bool:
        # save image
        cv2.imwrite(OutputFolder_DimensionImage_Path_str + fileName, Dimensions_Image)

    # estimate fruit area
    Area_Image = np.zeros((GrayScale_Image.shape[0], GrayScale_Image.shape[1], 1), np.uint8)
    Isolate_Image = np.zeros((Original_Image.shape[0], Original_Image.shape[1], 3), np.uint8)
    # throughout the pixel image to check it belongs inside contour or not
    for x in range (0, M):
        for y in range (0, N):
            # using pointPolygonTest to testing if a point is inside contour.
            # the return values are simply +1, -1, or 0 
            # depending on whether the point is inside, out-side, 
            # or on an edge.
            if (cv2.pointPolygonTest(Contours_Sorting_MaxMin[Fruit_Contours_SortingMaxMin_Idx_int], (y, x), False) > 0):
                Area_Image[x, y] = L - 1
                Isolate_Image[x, y] = Original_Image[x, y]
    if SaveImageProcessingResult_Check_bool:
        # save image
        cv2.imwrite(OutputFolder_AreaImage_Path_str + fileName, Area_Image)
        cv2.imwrite(OutputFolder_IsolateImage_Path_str + fileName, Isolate_Image)
    

InputFolder_Path_str = input("[ENTER]: Input directory (../dataset/mango/): ") # ../dataset/mango/
print("[INFOR]: Input directory: {}".format(InputFolder_Path_str))
OutputFolder_Path_str = input("[ENTER]: Output directory (../image_result/): ") # ../image_result/
print("[INFOR]: Output directory: {}".format(OutputFolder_Path_str))

# check whether using the multi threading or not
MultiThreading_Check_str = input("[ENTER]: Multi threading? y/n: ")
if MultiThreading_Check_str == "y" or MultiThreading_Check_str == "Y" or\
    MultiThreading_Check_str == "yes" or MultiThreading_Check_str == "Yes":
    MultiThreading_Check_bool = True
    print("[INFOR]: Multi threading: {}".format(MultiThreading_Check_bool))
elif MultiThreading_Check_str == "n" or MultiThreading_Check_str == "N" or\
    MultiThreading_Check_str == "no" or MultiThreading_Check_str == "No":
    MultiThreading_Check_bool = False
    print("[INFOR]: Multi threading: {}".format(MultiThreading_Check_bool))
else:
    MultiThreading_Check_bool = False
    print("[INFOR]: Multi threading: (Default) - {}".format(MultiThreading_Check_bool))

# choose image processing method
ImageProcessing_Method_str = input('''[INFOR]: Choosing image processing method:
[INFOR]:    Convert to HSV channel: 1
[INFOR]:    InRange: 2
[INFOR]:    Threshold: 3
[INFOR]:    Default (NOT convert to HSV channel, inrange and threshold): 0
[ENTER]: (0, 1, 2, or 3): ''')
if ImageProcessing_Method_str == "1":
    ImageProcessing_Method_str = "hsv"
    print("[INFOR]: Image processing method: {}".format(ImageProcessing_Method_str))
elif ImageProcessing_Method_str == "2":
    ImageProcessing_Method_str = "inrange"
    print("[INFOR]: Image processing method: {}".format(ImageProcessing_Method_str))
elif ImageProcessing_Method_str == "3":
    ImageProcessing_Method_str = "threshold"
    print("[INFOR]: Image processing method: {}".format(ImageProcessing_Method_str))
elif ImageProcessing_Method_str == "0":
    ImageProcessing_Method_str = "default"
    print("[INFOR]: Image processing method: {}".format(ImageProcessing_Method_str))
else:
    ImageProcessing_Method_str = "default"
    print("[INFOR]: Image processing method: {}".format(ImageProcessing_Method_str))

# check whether save image processing result or not
SaveImageProcessingResult_Check_str = input("[ENTER] Save image processing result? y/n: ")
if SaveImageProcessingResult_Check_str == "y" or SaveImageProcessingResult_Check_str == "Y" or\
    SaveImageProcessingResult_Check_str == "yes" or SaveImageProcessingResult_Check_str == "Yes":
    SaveImageProcessingResult_Check_bool = True
    print("[INFOR]: Save image processing result: {}".format(SaveImageProcessingResult_Check_bool))
elif SaveImageProcessingResult_Check_str == "n" or SaveImageProcessingResult_Check_str == "N" or\
    SaveImageProcessingResult_Check_str == "no" or SaveImageProcessingResult_Check_str == "No":
    SaveImageProcessingResult_Check_bool = False
    print("[INFOR]: Save image processing result: {}".format(SaveImageProcessingResult_Check_bool))
else:
    SaveImageProcessingResult_Check_bool = False
    print("[INFOR]: Save image processing result: (Default) - {}".format(SaveImageProcessingResult_Check_bool))

# the number of possible intensity levels in the image 
# (256 for an 8-bit image).
L = 256

# initialise the kernel size of 'blur' function.
"""
it means each pixel in the output is the simple mean
of all of the pixels in a window (the kernel),
around the corresponding pixel in the input.
"""
Blur_KernelSize_tuple = (5, 5)
print("[INFOR]: Blur kernel size: " + str(Blur_KernelSize_tuple))

# inRange
if ImageProcessing_Method_str == "inrange":
    inRange1_int = 30
    inRange2_int = 250
    print("[INFOR]: inRange1: " + str(inRange1_int))
    print("[INFOR]: inRange2: " + str(inRange2_int))

# initialise two thresholds, a lower (ThresholdCanny1_int) 
# and an upper (ThresholdCanny2_int) of canny edge detector.
"""
If a pixel has a gradient larger than the upper threshold,
then it is accepted as an edge pixel; if a pixel is below
the lower threshold, it is rejected. 
If the pixel’s gradient is between the thresholds, 
then it will be accepted only if it is connected to 
a pixel that is above the high threshold.
NOTE: Canny recommended a ratio of high:low threshold
        between 2:1 and 3:1
"""
ThresholdCanny1_int = 0
ThresholdCanny2_int = 3 * ThresholdCanny1_int
print("[INFOR]: Threshold Canny 1: " + str(ThresholdCanny1_int))
print("[INFOR]: Threshold Canny 2: " + str(ThresholdCanny2_int))

# initialise the kernel of dilation function
"""
In dilation function, any given pixel is replaced with
the local maximum of all of the pixel values covered by
the kernel.
"""
Dilate_KernelSize_np = np.ones((3, 3), np.uint8)
print("[INFOR]: Dilate Kernel Size: ", Dilate_KernelSize_np)

# flood fill
Gray_FloodFill_Ini_int = 10
print("[INFOR]: Gray FloodFill: " + str(Gray_FloodFill_Ini_int))
Gray_FloodFill_Step_int = 1
print("[INFOR]: Gray FloodFill Step: " + str(Gray_FloodFill_Step_int))
FloodFill_AreaComponents_MaxMin_Idx0_int = 0
print("[INFOR]: FloodFill Area Components MaxMin Idx 0: " + str(FloodFill_AreaComponents_MaxMin_Idx0_int))
FloodFill_AreaComponents_MaxMin_Idx1_int = 0
print("[INFOR]: FloodFill Area Components MaxMin Idx 1: " + str(FloodFill_AreaComponents_MaxMin_Idx1_int))

# contour
Fruit_Contours_SortingMaxMin_Idx_int = 0
print("Fruit Contours Sorting MaxMin Idx: " + str(Fruit_Contours_SortingMaxMin_Idx_int))

InputFileNames_list = os.listdir(InputFolder_Path_str)

if SaveImageProcessingResult_Check_bool == True:
    OutputFolder_GrayImage_Path_str = OutputFolder_Path_str + "gray/"
    if not os.path.exists(OutputFolder_GrayImage_Path_str):
        os.makedirs(OutputFolder_GrayImage_Path_str)
    OutputFolder_BlurImage_Path_str = OutputFolder_Path_str + "blur/"
    if not os.path.exists(OutputFolder_BlurImage_Path_str):
        os.makedirs(OutputFolder_BlurImage_Path_str)
    if ImageProcessing_Method_str == "inrange":
        OutputFolder_InrangeImage_Path_str = OutputFolder_Path_str + "inrange/"
        if not os.path.exists(OutputFolder_InrangeImage_Path_str):
            os.makedirs(OutputFolder_InrangeImage_Path_str)
    elif ImageProcessing_Method_str == "hsv":
        OutputFolder_HsvImage_Path_str = OutputFolder_Path_str + "hsv/"
        if not os.path.exists(OutputFolder_HsvImage_Path_str):
            os.makedirs(OutputFolder_HsvImage_Path_str)
    OutputFolder_CannyImage_Path_str = OutputFolder_Path_str + "canny/"
    if not os.path.exists(OutputFolder_CannyImage_Path_str):
        os.makedirs(OutputFolder_CannyImage_Path_str)
    OutputFolder_DilateImage_Path_str = OutputFolder_Path_str + "dilate/"
    if not os.path.exists(OutputFolder_DilateImage_Path_str):
        os.makedirs(OutputFolder_DilateImage_Path_str)
    OutputFolder_FloodFillImage_Path_str = OutputFolder_Path_str + "floodfill/"
    if not os.path.exists(OutputFolder_FloodFillImage_Path_str):
        os.makedirs(OutputFolder_FloodFillImage_Path_str)
    OutputFolder_RemoveNoiseImage_Path_str = OutputFolder_Path_str + "remove_noise/"
    if not os.path.exists(OutputFolder_RemoveNoiseImage_Path_str):
        os.makedirs(OutputFolder_RemoveNoiseImage_Path_str)
    OutputFolder_ContourImage_Path_str = OutputFolder_Path_str + "contour/"
    if not os.path.exists(OutputFolder_ContourImage_Path_str):
        os.makedirs(OutputFolder_ContourImage_Path_str)
    OutputFolder_AreaImage_Path_str = OutputFolder_Path_str + "area/"
    if not os.path.exists(OutputFolder_AreaImage_Path_str):
        os.makedirs(OutputFolder_AreaImage_Path_str)
    OutputFolder_DimensionImage_Path_str = OutputFolder_Path_str + "dimension/"
    if not os.path.exists(OutputFolder_DimensionImage_Path_str):
        os.makedirs(OutputFolder_DimensionImage_Path_str)
    OutputFolder_IsolateImage_Path_str = OutputFolder_Path_str + "isolate/"
    if not os.path.exists(OutputFolder_IsolateImage_Path_str):
        os.makedirs(OutputFolder_IsolateImage_Path_str)

print("------------START-PROCESSING------------")

if MultiThreading_Check_bool == True:
    StartTimeTotal = time.time()
    multiThread(InputFileNames_list)
    print("[INFOR]: Total time processing: {}".format(time.time() - StartTimeTotal))
    print("---------------DONE---------------")
else:
    StartTimeTotal = time.time()
    for InputFileName_Idx, InputFileName in enumerate(InputFileNames_list):
        print('File name:\t{}\t{}/{}'.format(InputFileName, InputFileName_Idx + 1, len(InputFileNames_list)))
        fruits_feature_extraction(InputFileName)

    print("[INFOR]: Total time processing: {}".format(time.time() - StartTimeTotal))
    print("---------------DONE---------------")

File: C_C++/Fruits_Segmentation.cpp
#include "Fruits_Segmentation.h"

using namespace cv;

int M;
int N;

char InputFolder_Path[100];
char OutputFolder_Path[100];

string OutputFolder_GrayImage_Path;
string OutputFolder_BlurImage_Path;
string OutputFolder_InrangeImage_Path;
string OutputFolder_CannyImage_Path;
string OutputFolder_DilateImage_Path;
string OutputFolder_FloodFillImage_Path;
string OutputFolder_RemoveNoiseImage_Path;
string OutputFolder_ContourImage_Path;

vector<string> vFileNames;
int FileNames_index;
//vector<string> vPathFileNames;

int Num_FileNames;
int Num_PathFileNames;

int check_mkdir;

Mat Original_Image;
Mat GrayScale_Image;
Mat Blur_Image;
Mat inRange_Image;
Mat Canny_Image;
Mat Dilate_Image;
Mat FloodFill_Image;
Mat RemoveNoise_Image;
Mat DrawFruitContour_Image;

int area[MaxIntensity];
int max_area;
int rmax_area;

// initialize before using 'findContours' function
vector<vector<Point>> contours;
vector<Vec4i> hierarchy;
int num_contours;

int largest_contour_value;
int largest_contour_index;

void Fruits_Segmentation(void)
{
	printf("[INFOR]: FRUITS SEGMENTATION C/C++\n");

	GetInput();

	cout << "----------START-PROCESSING----------" << endl;

	time_t totalTime_Start;
	time_t totalTime_End;

	time(&totalTime_Start);
	for (FileNames_index = 1; FileNames_index <= Num_FileNames; FileNames_index++)
	{
		printf("[INFOR]:\t%s\t%d/%d\n", vFileNames[FileNames_index].c_str(), FileNames_index, Num_FileNames);
		Remove_Noise(InputFolder_Path + vFileNames[FileNames_index]);
	}
	time(&totalTime_End);

	double totalTime_Taken = double(totalTime_End - totalTime_Start);

	printf("[INFOR]: Total time execution: %f\n", totalTime_Taken);
	cout << "----------------DONE----------------" << endl;

	return;
}

void GetInput(void)
{
	// enter input directory
	printf("[ENTER]: Input directory (../dataset/mango/): ");
	scanf("%s", InputFolder_Path);
	printf("[INFOR]: Input directory: %s\n", InputFolder_Path);
	if (InputFolder_Path[strlen(InputFolder_Path) - 1] != '*')
	{
		strcat(InputFolder_Path, "*"); // append '*'
	}

	// get all file names of input directory
	vFileNames = GetFileNamesInDirectory(InputFolder_Path);
	Num_FileNames = int(vFileNames.size()) - 1;		// casting to int to remove warning:
													// warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data.
	//printf("Num_FileNames: %d\n", Num_FileNames);
	//for (int i = 1; i < Num_FileNames + 1; i++) 
	//{
	//	cout << vFileNames[i] << endl;
	//	//printf("%s\n", vFileNames[i]);
	//}

	InputFolder_Path[strlen(InputFolder_Path) - 1] = '\0'; // remove last char (*)

	//// combine input folder path with file name
	//for (int i = 0; i < Num_FileNames; i++)
	//{
	//	vPathFileNames.push_back(InputFolder_Path + vFileNames[i + 1]);
	//}
	//Num_PathFileNames = int(vPathFileNames.size());	// casting to int to remove warning:
	//												// warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data.
	////printf("Num_PathFileNames: %d\n", Num_PathFileNames);
	////for (int i = 0; i < Num_PathFileNames; i++)
	////{
	////	cout << vPathFileNames[i] << endl;
	////	//printf("%s\n", vPathFileNames[i]);
	////}

	// enter output directory
	printf("[ENTER]: Output directory (../image_result/): ");
	scanf("%s", OutputFolder_Path);
	printf("[INFOR]: Output directory: %s\n", OutputFolder_Path);

	OutputFolder_GrayImage_Path = OutputFolder_Path;
	OutputFolder_GrayImage_Path = OutputFolder_GrayImage_Path.append("gray/");			// append 'gray'
	// create directory
	check_mkdir = _mkdir(OutputFolder_GrayImage_Path.c_str());
	if (!check_mkdir)
		printf("[INFOR]: The creation of GRAY image directory was: SUCCESSFUL.\n");
	else
		printf("[INFOR]: The creation of GRAY image directory was: UNSUCCESSFUL.\n");

	OutputFolder_BlurImage_Path = OutputFolder_Path;
	OutputFolder_BlurImage_Path = OutputFolder_BlurImage_Path.append("blur/");			// append 'blur'
	// create directory
	check_mkdir = _mkdir(OutputFolder_BlurImage_Path.c_str());
	if (!check_mkdir)
		printf("[INFOR]: The creation of BLUR image directory was: SUCCESSFUL.\n");
	else
		printf("[INFOR]: The creation of BLUR image directory was: UNSUCCESSFUL.\n");

	OutputFolder_InrangeImage_Path = OutputFolder_Path;
	OutputFolder_InrangeImage_Path = OutputFolder_InrangeImage_Path.append("inrange/");	// append 'inrange'
	// create directory
	check_mkdir = _mkdir(OutputFolder_InrangeImage_Path.c_str());
	if (!check_mkdir)
		printf("[INFOR]: The creation of INRANGE image directory was: SUCCESSFUL.\n");
	else
		printf("[INFOR]: The creation of INRANGE image directory was: UNSUCCESSFUL.\n");

	OutputFolder_CannyImage_Path = OutputFolder_Path;
	OutputFolder_CannyImage_Path = OutputFolder_CannyImage_Path.append("canny/");		// append 'canny'
	// create directory
	check_mkdir = _mkdir(OutputFolder_CannyImage_Path.c_str());
	if (!check_mkdir)
		printf("[INFOR]: The creation of CANNY image directory was: SUCCESSFUL.\n");
	else
		printf("[INFOR]: The creation of CANNY image directory was: UNSUCCESSFUL.\n");

	OutputFolder_DilateImage_Path = OutputFolder_Path;
	OutputFolder_DilateImage_Path = OutputFolder_DilateImage_Path.append("dilate/");	// append 'dilate'
	// create directory
	check_mkdir = _mkdir(OutputFolder_DilateImage_Path.c_str());
	if (!check_mkdir)
		printf("[INFOR]: The creation of DILATE image directory was: SUCCESSFUL.\n");
	else
		printf("[INFOR]: The creation of DILATE image directory was: UNSUCCESSFUL.\n");

	OutputFolder_FloodFillImage_Path = OutputFolder_Path;
	OutputFolder_FloodFillImage_Path = OutputFolder_FloodFillImage_Path.append("floodfill/"); // append 'floodfill'
	// create directory
	check_mkdir = _mkdir(OutputFolder_FloodFillImage_Path.c_str());
	if (!check_mkdir)
		printf("[INFOR]: The creation of FLOODFILL image directory was: SUCCESSFUL.\n");
	else
		printf("[INFOR]: The creation of FLOODFILL image directory was: UNSUCCESSFUL.\n");

	OutputFolder_RemoveNoiseImage_Path = OutputFolder_Path;
	OutputFolder_RemoveNoiseImage_Path = OutputFolder_RemoveNoiseImage_Path.append("remove_noise/"); // append 'remove_noise'
	// create directory
	check_mkdir = _mkdir(OutputFolder_RemoveNoiseImage_Path.c_str());
	if (!check_mkdir)
		printf("[INFOR]: The creation of REMOVE_NOISE image directory was: SUCCESSFUL.\n");
	else
		printf("[INFOR]: The creation of REMOVE_NOISE image directory was: UNSUCCESSFUL.\n");

	OutputFolder_ContourImage_Path = OutputFolder_Path;
	OutputFolder_ContourImage_Path = OutputFolder_ContourImage_Path.append("contour/");	// append 'contour'
	// create directory
	check_mkdir = _mkdir(OutputFolder_ContourImage_Path.c_str());
	if (!check_mkdir)
		printf("[INFOR]: The creation of CONTOUR image directory was: SUCCESSFUL.\n");
	else
		printf("[INFOR]: The creation of CONTOUR image directory was: UNSUCCESSFUL.\n");

	/*strcat(OutputFolder_BlurImage_Path, OutputFolder_Path);
	strcat(OutputFolder_BlurImage_Path, "blur/");						// append 'blur'

	strcat(OutputFolder_InrangeImage_Path, OutputFolder_Path);
	strcat(OutputFolder_InrangeImage_Path, "inrange/");					// append 'inrange'

	strcat(OutputFolder_CannyImage_Path, OutputFolder_Path);
	strcat(OutputFolder_CannyImage_Path, "canny/");						// append 'canny'

	strcat(OutputFolder_DilateImage_Path, OutputFolder_Path);
	strcat(OutputFolder_DilateImage_Path, "dilate/");					// append 'dilate'

	strcat(OutputFolder_FloodFillImage_Path, OutputFolder_Path);
	strcat(OutputFolder_FloodFillImage_Path, "floodfill/");				// append 'floodfill'

	strcat(OutputFolder_RemoveNoiseImage_Path, OutputFolder_Path);
	strcat(OutputFolder_RemoveNoiseImage_Path, "remove_noise/");		// append 'remove_noise'

	strcat(OutputFolder_ContourImage_Path, OutputFolder_Path);
	strcat(OutputFolder_ContourImage_Path, "contour/");					// append 'contour'
	cout << OutputFolder_ContourImage_Path << endl;
	printf("%s\n", OutputFolder_ContourImage_Path);*/
	
	return;
}

vector<string> GetFileNamesInDirectory(string directory) 
{
	vector<string> files;
	WIN32_FIND_DATA fileData;
	HANDLE hFind;
	if (!((hFind = FindFirstFile(directory.c_str(), &fileData)) == INVALID_HANDLE_VALUE)) 
	{				
		while (FindNextFile(hFind, &fileData)) 
		{		
			files.push_back(fileData.cFileName);
		}
	}
	FindClose(hFind);
	return files;
}

void Remove_Noise(string InputImagePath)
{
	// load image
	Original_Image = imread(InputImagePath);
	/*namedWindow("Original_Image", CV_WINDOW_AUTOSIZE);
	imshow("Original_Image", Original_Image);*/

	// convert to grayscale
	cvtColor(Original_Image, GrayScale_Image, COLOR_RGB2GRAY);
	/*namedWindow("GrayScale_Image", CV_WINDOW_AUTOSIZE);
	imshow("GrayScale_Image", GrayScale_Image);*/
	imwrite(OutputFolder_GrayImage_Path + vFileNames[FileNames_index], GrayScale_Image);

	M = GrayScale_Image.size().height;	// height of image
	N = GrayScale_Image.size().width;	// width of image

	// filter image and reduce noise of background
	blur(GrayScale_Image, Blur_Image, Size(5, 5));
	/*namedWindow("Blur_Image", CV_WINDOW_AUTOSIZE);
	imshow("Blur_Image", Blur_Image);*/
	imwrite(OutputFolder_BlurImage_Path + vFileNames[FileNames_index], Blur_Image);

	// inRange
	inRange(Blur_Image, 30, 250, inRange_Image);
	/*namedWindow("inRange_Image", CV_WINDOW_AUTOSIZE);
	imshow("inRange_Image", inRange_Image);*/
	imwrite(OutputFolder_InrangeImage_Path + vFileNames[FileNames_index], inRange_Image);

	// using canny algorithm to find out the edge
	Canny(inRange_Image, Canny_Image, 0, 3 * 0);
	/*namedWindow("Canny_Image", CV_WINDOW_AUTOSIZE);
	imshow("Canny_Image", Canny_Image);*/
	imwrite(OutputFolder_CannyImage_Path + vFileNames[FileNames_index], Canny_Image);

	// using dilate operation to try to find connected edge components
	Mat w = getStructuringElement(MORPH_RECT, Size(3, 3));
	dilate(Canny_Image, Dilate_Image, w);
	//namedWindow("Dilate_Image", CV_WINDOW_AUTOSIZE);
	//imshow("Dilate_Image", Dilate_Image);
	imwrite(OutputFolder_DilateImage_Path + vFileNames[FileNames_index], Dilate_Image);

	int gray = 10;
	int x, y;
	int r;
	FloodFill_Image = Dilate_Image.clone();
	// throughout each pixel value of the image
	for (x = 0; x < M; x++)
		for (y = 0; y < N; y++)
		{
			// take the current pixel value
			r = FloodFill_Image.at<uchar>(x, y);
			// check white pixel value
			if (r == MaxIntensity - 1)
			{
				// using 'floodFill' to mark or isolate portions of image
				floodFill(FloodFill_Image, Point(y, x), CV_RGB(gray, gray, gray));
				gray++;
			}
		}
	/*namedWindow("FloodFill_Image", CV_WINDOW_AUTOSIZE);
	imshow("FloodFill_Image", FloodFill_Image);*/
	imwrite(OutputFolder_FloodFillImage_Path + vFileNames[FileNames_index], FloodFill_Image);

	// determine area (the number of pixels) of each portions of image
	for (r = 0; r < MaxIntensity; r++)
		area[r] = 0;
	for (x = 0; x < M; x++)
		for (y = 0; y < N; y++)
		{
			r = FloodFill_Image.at<uchar>(x, y);
			if (r > 0) // the pixel value of each portions of image is now equal or above 'GRAY_FLOODFILL'
				area[r]++;
		}
	max_area, rmax_area = maxElementIndex(area, MaxIntensity);
	// remove noise
	RemoveNoise_Image = Mat(FloodFill_Image.size(), CV_8UC1);
	for (x = 0; x < M; x++)
		for (y = 0; y < N; y++)
		{
			r = FloodFill_Image.at<uchar>(x, y);
			if (r == rmax_area)
				RemoveNoise_Image.at<uchar>(x, y) = MaxIntensity - 1;
			else
				RemoveNoise_Image.at<uchar>(x, y) = 0;
		}
	/*namedWindow("RemoveNoise_Image", CV_WINDOW_AUTOSIZE);
	imshow("RemoveNoise_Image", RemoveNoise_Image);*/
	imwrite(OutputFolder_RemoveNoiseImage_Path + vFileNames[FileNames_index], RemoveNoise_Image);

	// create a black background image to draw contours on it
	DrawFruitContour_Image = Mat::zeros(RemoveNoise_Image.size(), CV_8UC3);
	// find contours
	findContours(RemoveNoise_Image, contours, hierarchy, CV_RETR_EXTERNAL, CV_CHAIN_APPROX_SIMPLE, Point(0, 0));
	// take the number of contours
	num_contours = (int)(contours.size());	// casting to int to remove warning:
											// warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data.
	// find the largest contour
	largest_contour_value = int(contours[0].size());	// casting to int to remove warning:
														// warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data.
	largest_contour_index = 0;
	for (int i = 1; i < num_contours; i++)
	{
		if (contours[i].size() > largest_contour_value)
		{
			largest_contour_value = int(contours[i].size());	// casting to int to remove warning:
																// warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data.

			largest_contour_index = i;
		}
	}
	// draw the largest contour
	drawContours(DrawFruitContour_Image, contours, largest_contour_index, Scalar(0, 255, 0), 1, 8, hierarchy, 0, Point());
	/*namedWindow("DrawFruitContour_Image", CV_WINDOW_AUTOSIZE);
	imshow("DrawFruitContour_Image", DrawFruitContour_Image);*/
	imwrite(OutputFolder_ContourImage_Path + vFileNames[FileNames_index], DrawFruitContour_Image);

	return;
}